### PR TITLE
Add dynamic parsing module

### LINKED ISSUE #3: Dynamically generate struct
The model struct is currently manually defined:

https://github.com/agate-model/Agate.jl/blob/dynamic_parsing/src/Models/NPZD/Static/struct_static.jl
(and):
https://github.com/agate-model/Agate.jl/blob/dynamic_parsing/src/Models/NPZD/Dynamic/struct_dynamic.jl

This needs to be dynamically generated based on the YAML specification instead.

e.g.:
https://github.com/agate-model/Agate.jl/blob/dynamic_parsing/src/Models/NPZD/Dynamic/NPZD.yml

### LINKED ISSUE #4: Dynamically generate function
The model function is currently manually defined:

https://github.com/agate-model/Agate.jl/blob/dynamic_parsing/src/Models/NPZD/Static/function_static.jl
(and):
https://github.com/agate-model/Agate.jl/blob/dynamic_parsing/src/Models/NPZD/Dynamic/function_dynamic.jl

This needs to be dynamically generated based on the YAML specification instead.

e.g.:
https://github.com/agate-model/Agate.jl/blob/dynamic_parsing/src/Models/NPZD/Dynamic/NPZD.yml

### LINKED ISSUE #6: Dynamic Parsing and other functions should be their own modules
Dynamic parsing is currently in NPZD/dynamic:

https://github.com/agate-model/Agate.jl/blob/dynamic_parsing/src/Models/NPZD/Dynamic/dynamic_parsing.jl

This function, and other functions (e.g. https://github.com/agate-model/Agate.jl/issues/3, and https://github.com/agate-model/Agate.jl/issues/4) should be wrapped in a Module such that they can be easily called.

Something like this;

```
using Agate
my_model = parse_yaml("path_to_yaml")
```

### LINKED ISSUE #7: Add dynamic parsing checks
Add checks to make sure functions defined in YAML are found in the library, and all parameter values are provided.

### PR DESCRIPTION
**Pull request by radka-j**
Closes #3 
Closes #4 
Closes #6 
Closes #7 

## Overview

This PR adds a Dynamic module with two functions, `create_bgc_struct` and `add_bgc_methods`. This enables the following workflow:

```Julia
using Agate

parameters = (α=2/3, β=4/3,  δ=1, γ=1) #NamedTuple or Dict
tracers = Dict("R" => :(α*R - β*R*F), "F" => :(-γ*F + δ*R*F))

# the create_struct method creates an Agate.Models.Dynamic.LV object so here
# we assign it to a local LV variable - a bit annoying to have to do this
LV = create_bgc_struct(:LV, parameters)
add_bgc_methods(LV, tracers) 

# the model is instantiated with the default parameter values unless other values are provided here
model = LV() 
```

The model object can then be passed to Oceananigans.

There is an option for the user to pass functions to `add_bgc_methods` which are used within the tracer expressions (these could be imported from other libraries). The tests show an example of the NPZD model implemented in this way.

## For reviewers

The key question is: is this the kind of interface we imagined? Can anything be improved? Are there any potential misuses/bugs we can imagine the user making? 

Otherwise, any comments on implementation and code style are always more than welcome.

Note that there is some additional complexity (e.g., light, sinking particles, boundary conditions) that will be addressed in future PRs (see #10 for discussion).  In its current form the returned model would need to be passed to OceanBioME [Biogeochemistry](https://github.com/OceanBioME/OceanBioME.jl/blob/main/src/OceanBioME.jl) to make it fully Oceananigans compatible.

### CODE REVIEW

**Review by nanophyto — COMMENTED** [review]
Hi @radka-j many thanks for putting this together. These are my comments thus far, I'll have a further look this afternoon. All minor stuff currently, I'll test some edge cases with the model formulation later.

**Comment by nanophyto** [comment]
> This PR adds a Dynamic module with two functions, `create_bgc_struct` and `add_bgc_methods`. This enables the following workflow:
> 
> ```julia
> using Agate
> 
> parameters = (α=2/3, β=4/3,  δ=1, γ=1) #NamedTuple or Dict
> tracers = Dict("R" => :(α*R - β*R*F), "F" => :(-γ*F + δ*R*F))
> 
> # the create_struct method creates an Agate.Models.Dynamic.LV object so here
> # we assign it to a local LV variable - a bit annoying to have to do this
> LV = create_bgc_struct(:LV, parameters)
> add_bgc_methods(LV, tracers) 
> 
> # the model is instantiated with the default parameter values unless other values are provided here
> model = LV() 
> ```
> 
> The model object can then be passed to Oceananigans.


It might be useful to include this as an example as part of the PR? e.g. in `examples/`

File: src/Models/Dynamic.jl [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            function (bgc::$(bgc_type))(::Val{Symbol($tracer_name)}, $(all_state_vars...))
+                $(method_vars...)
+                return $(tracer_expression)
+            end
+        end
+        eval(tracer_method)
+    end
+
+    # Q: this return statement is unnecessary - is it good practice to leave or remove it?
+    return bgc_type
```
Comment (ljwolf) [inline src/Models/Dynamic.jl:131]: I'd always return, since it makes future method chaining easy.

File: test/test_dynamic.jl [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        f_expr = :(f1(x, α) + f2(y, β))
+        params = [:α, :β, :x, :y]
+        @test_throws UndefVarError expression_check(params, f_expr)
+
+        # no errors - using base methods
+        f_expr = :(α * x - β * x * y)
+        params = [:α, :β, :x, :y]
+        @test expression_check(params, f_expr) === nothing
+
+        # no erors - method defined in Dynamic module
```
Comment (ljwolf) [inline test/test_dynamic.jl:25]:
```suggestion
        # no errors - method defined in Dynamic module
```

File: test/test_dynamic.jl [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            @test model2(Val(:F), 0, 0, 0, 0, 10, 2, 0) == -2 * 2 + 2 * 10 * 2
+
+        end
+
+        @testset "use helper functions" begin
+
+            # NPZD model
+
+            # Q: should this be a temp file?
+            helper_functions = "./fixtures/NPZD.jl"
```
Comment (ljwolf) [inline test/test_dynamic.jl:92]: I don't think so. We will probably reuse the fixture over other tests. That said, should this be pulled from the example folder instead of put within the tests? In that case, the setup phase of the CI runner should copy the examples into the test locations as temp files.

It's great to use examples as part of the CI, but then keeping the tested examples in sync with the public examples can be annoying.

File: src/Models/Dynamic.jl [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+"""
+    parse_expression(f_expr) -> Tuple(Vector, Vector)
+
+Returns all argument names and methods called in expression.
+
+# Example
+```Julia
+parse_expression(:(α * x - β * x * y))
+"""
+function parse_expression(f_expr)
```
Comment (ljwolf) [inline src/Models/Dynamic.jl:144]: Still very neat 😁

File: src/Models/Dynamic.jl [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+# Arguments
+- `struct_name`: name for the new struct
+- `parameters`: named sequence of values of the form (field name = default value, ...)
+
+Note that the field names in parameters can't be any of [:x,:y,:z,:t].
+
+# Example
+create_bgc_struct(:LV, (α=2/3, β=4/3,  δ=1, γ=1))
+"""
+function create_bgc_struct(struct_name, parameters)
```
[the remark this answers is not in the document]
Comment (ljwolf) [reply src/Models/Dynamic.jl:31]: Would this be able to be solved by dispatch? It looks to me that the continuous and the discrete implementations differ in their input requirements. 

We can have one function but two methods with different signatures.

File: src/Models/Dynamic.jl [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+"""
+    add_bgc_methods(bgc_type, tracers, auxiliary_fields=[], helper_functions=()) -> DataType
+
+Add core methods to bgc_type required of AbstractContinuousFormBiogeochemistry:
+    - required_biogeochemical_tracers
+    - required_biogeochemical_auxiliary_fields
+    - a method per tracer
+
+# Arguments
+- `bgc_type`: subtype of AbstractContinuousFormBiogeochemistry
```
[the remark this answers is not in the document]
Comment (radka-j) [reply src/Models/Dynamic.jl:60]: Agree that it should be phrased better but I struggled to come up with what to change it to. Any suggestions?

File: src/Models/Dynamic.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (radka-j) [reply src/Models/Dynamic.jl:32]: Yes I think multiple dispatch could be a way to do it. I also agree with Joost that it's probably not high on the priority list. Should we make an issue for it?

File: test/test_dynamic.jl (hunk 3 above)
Comment (radka-j) [reply test/test_dynamic.jl:91]: Good question about whether this could be in examples. Another thing to think about is that these functions will eventually also be defined in the package somewhere whereas this test should  be for methods that can't accidentally be imported from within some module. Might be worth renaming the function names be less descriptive?

**Comment by nanophyto** [comment]
I quickly tested if the code works with nested functions:

e.g. the user might write something like this:

```
                "Z" => :(zooplankton_grazing_gain(P, Z, β, dummy_gmax_function(gₘₐₓ, a, b), kₚ)
                - zooplankton_metabolic_loss(Z, lᶻⁿ)
                - zooplankton_mortality_loss(Z, lᶻᵈ))
 ```


I updated NPZD.jl to include a dummy function which will always return the input value of gmax:

```dummy_gmax_function(gₘₐₓ, a, b) = ((gₘₐₓ*a) + b-b)/a # this will always return gₘₐₓ```

and then created a new test:

```

@testset "test nested functions" begin

            helper_functions = "./fixtures/NPZD.jl"

            parameters = (
                μ₀ = 0.6989,
                kₙ = 2.3868,
                lᵖⁿ = 0.066,
                lᶻⁿ = 0.0102,
                lᵖᵈ = 0.0101,
                gₘₐₓ = 2.1522,
                kₚ = 0.5573,
                β = 0.9116,
                lᶻᵈ = 0.3395,
                rᵈⁿ = 0.1213,
                α = 0.1953,
                a = 0.234,
                b = 9.2132
            )
            aux_field_vars = [:PAR]

            tracers = Dict(
                "N" => :(
                - phytoplankton_growth(N, PAR, P, μ₀, kₙ, α)
                + phytoplankton_metabolic_loss(P, lᵖⁿ)
                + zooplankton_metabolic_loss(Z, lᶻⁿ)
                + remineralization(D, rᵈⁿ)
                ),

                "D" => :(phytoplankton_mortality_loss(P, lᵖᵈ)
                + zooplankton_assimilation_loss(P, Z, β, gₘₐₓ, kₚ)
                + zooplankton_mortality_loss(Z, lᶻᵈ)
                - remineralization(D, rᵈⁿ)),

                "P" => :(phytoplankton_growth(N, PAR, P, μ₀, kₙ, α)
                - phytoplankton_grazing_loss(P, Z, gₘₐₓ, kₚ)
                - phytoplankton_metabolic_loss(P, lᵖⁿ)
                - phytoplankton_mortality_loss(P, lᵖᵈ)),

                "Z" => :(zooplankton_grazing_gain(P, Z, β, dummy_gmax_function(gₘₐₓ, a, b), kₚ)
                - zooplankton_metabolic_loss(Z, lᶻⁿ)
                - zooplankton_mortality_loss(Z, lᶻᵈ))
            )

            NPZD_nested = create_bgc_struct(:NPZD_nested, parameters)
            add_bgc_methods(NPZD_nested, tracers, auxiliary_fields=aux_field_vars, helper_functions=helper_functions)
            model_nested = NPZD_nested()

            Z = 0.05
            P = 0.01
            N = 7.0
            D = 1
            PAR = 1

            @test isapprox(model_nested(Val(:N), 0, 0, 0, 0, Z,P,N,D,PAR), 0.12101359881701153)
            @test isapprox(model_nested(Val(:D), 0, 0, 0, 0, Z,P,N,D,PAR), -0.12034718812821411)
            @test isapprox(model_nested(Val(:P), 0, 0, 0, 0, Z,P,N,D,PAR), 0.000660764624324505)
            @test isapprox(model_nested(Val(:Z), 0, 0, 0, 0, Z,P,N,D,PAR), -0.0013271753131219258)

        end

```
Which works great!

I don't think it is worth including the test above in `main` but it good to know it works.

**Comment by nanophyto** [comment]
OK, I think I found an edge case where things break.

If I introduce three new parameters:

```                
a = 0.6536,
b = 0.0633,
c = 1.4353
```
which sum to the value of gmax (2.1522)

And I then define something like this:
```
                "Z" => :(zooplankton_grazing_gain(P, Z, β, sum([a, b, c]), kₚ)
                - zooplankton_metabolic_loss(Z, lᶻⁿ)
                - zooplankton_mortality_loss(Z, lᶻᵈ))
 ```

I get the following error:

```
"UndefVarError(:a)", "UndefVarError: `a` not defined
```

It seems the issue is related to the variables being passed as an array?




````

        @testset "test base Julia functions" begin
            helper_functions = "./fixtures/NPZD.jl"

            parameters = (
                μ₀ = 0.6989,
                kₙ = 2.3868,
                lᵖⁿ = 0.066,
                lᶻⁿ = 0.0102,
                lᵖᵈ = 0.0101,
                gₘₐₓ = 2.1522,
                kₚ = 0.5573,
                β = 0.9116,
                lᶻᵈ = 0.3395,
                rᵈⁿ = 0.1213,
                α = 0.1953,
                a = 0.6536,
                b = 0.0633,
                c = 1.4353
            )
            aux_field_vars = [:PAR]

            tracers = Dict(
                "N" => :(
                - phytoplankton_growth(N, PAR, P, μ₀, kₙ, α)
                + phytoplankton_metabolic_loss(P, lᵖⁿ)
                + zooplankton_metabolic_loss(Z, lᶻⁿ)
                + remineralization(D, rᵈⁿ)
                ),

                "D" => :(phytoplankton_mortality_loss(P, lᵖᵈ)
                + zooplankton_assimilation_loss(P, Z, β, gₘₐₓ, kₚ)
                + zooplankton_mortality_loss(Z, lᶻᵈ)
                - remineralization(D, rᵈⁿ)),

                "P" => :(phytoplankton_growth(N, PAR, P, μ₀, kₙ, α)
                - phytoplankton_grazing_loss(P, Z, gₘₐₓ, kₚ)
                - phytoplankton_metabolic_loss(P, lᵖⁿ)
                - phytoplankton_mortality_loss(P, lᵖᵈ)),

                "Z" => :(zooplankton_grazing_gain(P, Z, β, dummy_sum([a, b, c]), kₚ)
                - zooplankton_metabolic_loss(Z, lᶻⁿ)
                - zooplankton_mortality_loss(Z, lᶻᵈ))
            )

            NPZD_sum = create_bgc_struct(:NPZD_sum, parameters)
            add_bgc_methods(NPZD_sum, tracers, auxiliary_fields=aux_field_vars, helper_functions=helper_functions)
            model_sum = NPZD_sum()

            Z = 0.05
            P = 0.01
            N = 7.0
            D = 1
            PAR = 1

            @test isapprox(model_sum(Val(:N), 0, 0, 0, 0, Z,P,N,D,PAR), 0.12101359881701153)
            @test isapprox(model_sum(Val(:D), 0, 0, 0, 0, Z,P,N,D,PAR), -0.12034718812821411)
            @test isapprox(model_sum(Val(:P), 0, 0, 0, 0, Z,P,N,D,PAR), 0.000660764624324505)
            @test isapprox(model_sum(Val(:Z), 0, 0, 0, 0, Z,P,N,D,PAR), -0.0013271753131219258)

        end
```

File: test/test_dynamic.jl (hunk 3 above)
Comment (ljwolf) [reply test/test_dynamic.jl:91]: We decided then to keep as many fixtures as possible within the example folder explicitly. All "examples" (or as many as possible) should be fixtures, and as few fixtures as possible should live separately in the `fixtures` folder.

**Review by ljwolf — APPROVED** [review]

File: Project.toml [hunk 7]
```diff
@@ -3,6 +3,9 @@ uuid = "41889b7c-c35c-4f16-abd7-94b299d9fd29"
 authors = ["Joost de Vries"]
```
[the remark this answers is not in the document]
Comment (ljwolf) [reply Project.toml:3]: We decided to fully remove the authors line: 
```suggestion

```

**Comment by radka-j** [comment]
The bug that @nanophyto found has to do with some simple assumptions I made about how the expressions are parsed that don't always hold up. This might require some careful rethinking (well spotted!!)

In all the examples we had until now, whenever I looked at the `args` of an expression, I always observed the same pattern where the first arg was the method name and the following args the parameters (which could be other expressions or values of symbols), e.g.:

```Julia
julia> :(α * x - β * x * y).args
3-element Vector{Any}:
 :-
 :(α * x)
 :(β * x * y)
```

Unfortunately array expressions don't have this pattern, the args are just the elements of the array as there are no methods:

```Julia
julia> :([a, b, c]).args
3-element Vector{Any}:
 :a
 :b
 :c
```

So in this case, the code I wrote interprets `a` as a method (which isn't defined anywhere) and this means it is missing from the parameters list.
